### ISSUE TITLE
Train/test procedure for link prediction on NE

### ISSUE THREAD

**Issue by akastrin** [Dru-Mara/EvalNE#6]
Dear authors,

I try to understand a procedure around train/test splitting for evaluating network embeddings on link prediction you implemented in your package. I posted a similar question yesterday on [Data Science Stack Exchange](https://datascience.stackexchange.com/questions/63341/link-prediction-on-network-embeddings).

If I understand your [paper](http://ceur-ws.org/Vol-2436/article_2.pdf) correctly, the node embedding is performed only on positive examples (edges) on the training set; false examples are used only to train a classifier? Is this correct?

Thanks.

Best, Andrej

**Comment by Dru-Mara** [comment]
Hi Andrej,

**The short answer is:**
You are correct, in order to compute node embeddings you only need the "true" train edges. These edges span a training graph which is the input we give NE methods to learn the node embeddings. Non-edges are only used in the binary classification problem.

**The longer answer is:**
The most effective way to compute link predictions from embeddings is via binary classification. The node embeddings first need to be transformed into edge embeddings (e.g. averaging the embeddings of _i_ and _j_ to get the embedding of _e=(i,j)_) and then solve a binary classification problem. The binary classifier (usually Logistic regression) will take the edge embeddings as examples and try to learn a mapping to their corresponding "label" where these labels are 1 for the "true" edges and 0 for the non-edges. Therefore, in addition to the edge embeddings of "true" graph edges, we need to compute the edge embeddings of a set of non-edges in the graphs (those with label 0). This is what we usually refer to when we talk about negative sampling for LP.

I believe the confusion here comes from the way some popular NE algorithms compute the embeddings. For instance, methods base on matrix factorization directly take the _train graph_ you give as input, compute the adjacency matrix, factorize it and return the node embeddings. Other methods e.g. Deepwalk, Node2vec or LINE have a special way of learning embeddings (Skip-gram from work embedding literature) and require "negative sampling" in the embedding learning process itself. These methods will select, from that train graph you give them, a set of non-edges or paths of non-connected edges to learn which node vectors should be put close together and which should be far. The NE methods will select these samples of non-edges themselves and in different ways, so the user generally doesn't have any control over it. 

I hope this clarifies your question,
Alex

**Comment by jlijffijt** [comment]
Hi Andrej,

I would like to add the following:

- If you want just to make predictions, all your data is training data and you can make predictions for data that you have not seen. You do not need to split your data into two parts.

- The splitting is hence only useful to evaluate the performance of link prediction methods. Your drawing on Stack Exchange does not reflect a standard machine learning pipeline. Unbiased estimates of performance can be computed only in the following manner: from the train data you construct a model (which here would be two parts; the network embedding and the binary classifier such as logistic regression), from which you can make predictions. You evaluate the predictions on the ground truth that was withheld from the learning process, which is the test set. Hence, the pipeline would be:

Left-hand side: Train network -> Network embedding -> LR model -> Predictions
Right-hand side: Test network -> Evaluation
Cross link from land-hand side ‘Predictions’ directly to right-hand side ‘Evaluation’.

So, the predictions are made from the model (for example logistic regression which is in turn based on a network embedding), which is learned on the training data. This is true not just in this link prediction setting but generally for any machine-learning model evaluation setting.

As an additional remark, there is no way to align network embeddings learned on the train and the test network, so an approach including a new embedding on the test set would not make sense.

Hope fully this clarifies the setting?

Best wishes,

~ Jef

On 19 Nov 2019, at 10:11, Dru-Mara <notifications@github.com<mailto:notifications@github.com>> wrote:


Hi Andrej,

The short answer is:
You are correct, in order to compute node embeddings you only need the "true" train edges. These edges span a training graph which is the input we give NE methods to learn the node embeddings. Non-edges are only used in the binary classification problem.

The longer answer is:
The most effective way to compute link predictions from embeddings is via binary classification. The node embeddings first need to be transformed into edge embeddings (e.g. averaging the embeddings of i and j to get the embedding of e=(i,j)) and then solve a binary classification problem. The binary classifier (usually Logistic regression) will take the edge embeddings as examples and try to learn a mapping to their corresponding "label" where these labels are 1 for the "true" edges and 0 for the non-edges. Therefore, in addition to the edge embeddings of "true" graph edges, we need to compute the edge embeddings of a set of non-edges in the graphs (those with label 0). This is what we usually refer to when we talk about negative sampling for LP.

I believe the confusion here comes from the way some popular NE algorithms compute the embeddings. For instance, methods base on matrix factorization directly take the train graph you give as input, compute the adjacency matrix, factorize it and return the node embeddings. Other methods e.g. Deepwalk, Node2vec or LINE have a special way of learning embeddings (Skip-gram from work embedding literature) and require "negative sampling" in the embedding learning process itself. These methods will select, from that train graph you give them, a set of non-edges or paths of non-connected edges to learn which node vectors should be put close together and which should be far. The NE methods will select these samples of non-edges themselves and in different ways, so the user generally doesn't have any control over it.

I hope this clarifies your question,
Alex

**Comment by akastrin** [comment]
Many thanks for both answers. However, @jlijffijt pointed out:

> So, the predictions are made from the model (for example logistic regression which is in turn based on a network embedding), which is learned on the training data.

So, just to clarify: edge representations (e.g., using Hadamard operator) in an evaluation phase are constructed from learned embeddings in the training phase?

Thanks again.

Best, Andrej

**Comment by Dru-Mara** [comment]
Yes, from the train network the NE method will generate a set of node embeddings and those are used to compute the edge embeddings of both train and test edges.

Alex
